File: server/core/module.go
package core

import (
	"fmt"
	"sync"

	"github.com/gorilla/mux"
)

type Module interface {
	RegisterRoutes(r *mux.Router)
	IsRunning() bool
	Enable(startNow bool) error
	Disable(stopNow bool) error
	Restart() error
	GetConfig(params map[string]string) (map[string]interface{}, error)
	SaveConfig(params map[string]string, config map[string]interface{}) error
}

var modules = make(map[string]Module)
var modulesMutex sync.RWMutex

func RegisterModule(name string, module Module) {
	modulesMutex.Lock()
	defer modulesMutex.Unlock()
	modules[name] = module
}

func GetModules() map[string]Module {
	modulesMutex.RLock()
	defer modulesMutex.RUnlock()
	var moduleList = map[string]Module{}
	for name, module := range modules {
		moduleList[name] = module
	}
	return moduleList
}

func GetModule(name string) Module {
	modulesMutex.RLock()
	defer modulesMutex.RUnlock()
	return modules[name]
}

func GetModuleStatus(name string) (bool, error) {
	modulesMutex.RLock()
	defer modulesMutex.RUnlock()
	if module, exists := modules[name]; exists {
		return module.IsRunning(), nil
	}
	return false, fmt.Errorf("module %s not found", name)
}

func EnableModule(name string, startNow bool) error {
	modulesMutex.RLock()
	defer modulesMutex.RUnlock()
	if module, exists := modules[name]; exists {
		return module.Enable(startNow)
	}
	return fmt.Errorf("module %s not found", name)
}

func DisableModule(name string, stopNow bool) error {
	modulesMutex.RLock()
	defer modulesMutex.RUnlock()
	if module, exists := modules[name]; exists {
		return module.Disable(stopNow)
	}
	return fmt.Errorf("module %s not found", name)
}

func RestartModule(name string) error {
	modulesMutex.RLock()
	defer modulesMutex.RUnlock()
	if module, exists := modules[name]; exists {
		return module.Restart()
	}
	return fmt.Errorf("module %s not found", name)
}

func GetModuleConfig(name string, params map[string]string) (map[string]interface{}, error) {
	modulesMutex.RLock()
	defer modulesMutex.RUnlock()
	if module, exists := modules[name]; exists {
		return module.GetConfig(params)
	}
	return nil, fmt.Errorf("module %s not found", name)
}

func SaveModuleConfig(name string, params map[string]string, config map[string]interface{}) error {
	modulesMutex.RLock()
	defer modulesMutex.RUnlock()
	if module, exists := modules[name]; exists {
		module.SaveConfig(params, config)
		return nil
	}

	return fmt.Errorf("module %s not found", name)
}

File: server/core/module.py
import threading

class Module:
    def register_routes(self, app):
        pass
    def is_running(self):
        return False
    def enable(self, start_now):
        pass
    def disable(self, stop_now):
        pass
    def restart(self):
        pass
    def get_config(self, params):
        return {}
    def save_config(self, params, config):
        pass

modules = {}
modules_mutex = threading.Lock()

def register_module(name, module):
    with modules_mutex:
        modules[name] = module

def get_modules():
    with modules_mutex:
        return dict(modules)

def get_module(name):
    with modules_mutex:
        return modules.get(name)

def get_module_status(name):
    with modules_mutex:
        module = modules.get(name)
        if module:
            return module.is_running(), None
    return False, Exception(f"module {name} not found")

def enable_module(name, start_now):
    with modules_mutex:
        module = modules.get(name)
    if module:
        try:
            module.enable(start_now)
            return None
        except Exception as e:
            return e
    return Exception(f"module {name} not found")

def disable_module(name, stop_now):
    with modules_mutex:
        module = modules.get(name)
    if module:
        try:
            module.disable(stop_now)
            return None
        except Exception as e:
            return e
    return Exception(f"module {name} not found")

def restart_module(name):
    with modules_mutex:
        module = modules.get(name)
    if module:
        try:
            module.restart()
            return None
        except Exception as e:
            return e
    return Exception(f"module {name} not found")

def get_module_config(name, params):
    with modules_mutex:
        module = modules.get(name)
    if module:
        try:
            return module.get_config(params), None
        except Exception as e:
            return None, e
    return None, Exception(f"module {name} not found")

def save_module_config(name, params, config):
    with modules_mutex:
        module = modules.get(name)
    if module:
        try:
            module.save_config(params, config)
            return None
        except Exception as e:
            return e
    return Exception(f"module {name} not found")
